### PR TITLE
Improve colocate table balance logic for backend added

### PR DESCRIPTION
**Pull request by kangkaisen**
For https://github.com/apache/incubator-doris/issues/1137

1. Abstract the balance logic for backend added to `balance` method.
2. Do clone job after colocate backendsPerBucketSeq meta update for both backend added or removed.

### CODE REVIEW

File: fe/src/main/java/org/apache/doris/clone/ColocateTableBalancer.java [hunk 1]
```diff
@@ -507,111 +508,133 @@ private Long selectCloneBackendIdForRemove(com.google.common.collect.Table<Long,
...
+                .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
 
-                            Long cloneReplicaBackendId = newGroup2BackendsPerBucketSeq.get(groupId, i);
-                            if (cloneReplicaBackendId == null) {
-                                // select dest backend
-                                cloneReplicaBackendId = addedBackendIds.get(i % addedBackendIds.size());
-                                newGroup2BackendsPerBucketSeq.put(groupId, i, cloneReplicaBackendId);
-                            }
+        return backendCounter.entrySet().stream().sorted(Collections.reverseOrder(Map.Entry.comparingByValue()))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, (e1, e2) -> e1, LinkedHashMap::new));
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/clone/ColocateTableBalancer.java]: Hash map is not ordered, I think you what you need is List<Map.Entry<Long, Long>>

**Review by imay — CHANGES_REQUESTED** [review]

File: fe/src/main/java/org/apache/doris/clone/ColocateTableBalancer.java [hunk 2]
```diff
@@ -507,111 +508,133 @@ private Long selectCloneBackendIdForRemove(com.google.common.collect.Table<Long,
...
+     * @return the balanced mapping from bucket seq to backend
+     */
+    public static List<List<Long>> balance(List<List<Long>> backendsPerBucketSeq, List<Long> newBackends) {
+        int replicateNum = backendsPerBucketSeq.get(0).size();
+        Set<Long> groupBackendSet = backendsPerBucketSeq.stream().flatMap(List::stream).collect(Collectors.toSet());
+        List<Long> groupBackendList = backendsPerBucketSeq.stream().flatMap(List::stream).collect(Collectors.toList());
+        Map<Long, Long> sortedBackendCounter = getBackendCounter(groupBackendList);
+
+        int allBackendSize = groupBackendSet.size() + newBackends.size();
+        //all backend should keep at least one replica
```
[the remark this answers is not in the document]
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/clone/ColocateTableBalancer.java]: done

File: fe/src/main/java/org/apache/doris/clone/ColocateTableBalancer.java [hunk 3]
```diff
@@ -507,111 +508,133 @@ private Long selectCloneBackendIdForRemove(com.google.common.collect.Table<Long,
...
+        int avgReplicaNum = Math.max(groupBackendList.size() / allBackendSize, 1);
+        //the most balance case: all replica in all bucket seq have different backend
+        int needBalanceNum = Math.min(avgReplicaNum * newBackends.size(), groupBackendList.size() - groupBackendSet.size());
```
[the remark this answers is not in the document]
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/clone/ColocateTableBalancer.java]: has comment

File: fe/src/main/java/org/apache/doris/clone/ColocateTableBalancer.java [hunk 4]
```diff
@@ -507,111 +508,133 @@ private Long selectCloneBackendIdForRemove(com.google.common.collect.Table<Long,
...
+        //keep which BucketSeq will migrate to the new target backend
+        Map<Long, List<Integer>> targetBackendsToBucketSeqs = Maps.newHashMap();
+
+        while (hasBalancedNum < needBalanceNum) {
+            for(Map.Entry<Long, Long> beckendCounter: sortedBackendCounter.entrySet()) {
+                long count = beckendCounter.getValue();
+                long sourceBackend = beckendCounter.getKey();
+
+                //new backend should not as sourceBackend
+                if (newBackends.contains(sourceBackend)) {
```
[the remark this answers is not in the document]
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/clone/ColocateTableBalancer.java]: has comment

File: fe/src/main/java/org/apache/doris/clone/ColocateTableBalancer.java [hunk 5]
```diff
@@ -507,111 +508,133 @@ private Long selectCloneBackendIdForRemove(com.google.common.collect.Table<Long,
...
+                    continue;
+                }
 
-                            //update ColocateTableIndex backendsPerBucketSeq
-                            List<Long> backends = newBackendsPerBucketSeq.get(i);
-                            backends.remove(deleteReplica.getBackendId());
-                            if (!backends.contains(cloneReplicaBackendId)) {
-                                backends.add(cloneReplicaBackendId);
-                            }
+                if (count > avgReplicaNum && hasBalancedNum < needBalanceNum) {
```
[the remark this answers is not in the document]
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/clone/ColocateTableBalancer.java]: I don't get this point.

File: fe/src/main/java/org/apache/doris/clone/ColocateTableBalancer.java [hunk 6]
```diff
@@ -507,111 +508,133 @@ private Long selectCloneBackendIdForRemove(com.google.common.collect.Table<Long,
...
+                    Long targetBackend = newBackends.get (hasBalancedNum % newBackends.size());
+
+                    List<Integer> targetIndexes = IntStream.range(0, groupBackendList.size()).boxed()
```
[the remark this answers is not in the document]
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/clone/ColocateTableBalancer.java]: done

File: fe/src/main/java/org/apache/doris/clone/ColocateTableBalancer.java [hunk 7]
```diff
@@ -507,111 +508,133 @@ private Long selectCloneBackendIdForRemove(com.google.common.collect.Table<Long,
...
+                            .filter(i -> groupBackendList.get(i).equals(sourceBackend))
+                            .collect(Collectors.toList());
+
+                    for(int targetIndex: targetIndexes) {
+                        int targetBucketSeq = targetIndex / replicateNum;
```
[the remark this answers is not in the document]
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/clone/ColocateTableBalancer.java]: done

File: fe/src/main/java/org/apache/doris/clone/ColocateTableBalancer.java (hunk 1 above)
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/clone/ColocateTableBalancer.java]: The input order and out order for `LinkedHashMap` is the same.

Comment (imay) [reply fe/src/main/java/org/apache/doris/clone/ColocateTableBalancer.java]: OK, I see

File: fe/src/main/java/org/apache/doris/clone/ColocateTableBalancer.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (imay) [reply fe/src/main/java/org/apache/doris/clone/ColocateTableBalancer.java]: Sorry for my mistake. what I want to express is when `hasBalancedNum ` >= `needBalanceNum ` we can break all the `while` loop.

[the remark this answers is not in the document]
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/clone/ColocateTableBalancer.java]: OK. I see. done.

**Review by imay — APPROVED** [review]
LGTM

**Pull request merged**